osf/management/commands/update_institution_sso_email_domain.py

```diff
@@ -0,0 +1,146 @@
+                update_result.get(UpdateResult.FAILED.name)[user._id] = user_result.get(UpdateResult.FAILED.name)
+            if user_result.get(UpdateResult.ERRORED.name):
+                update_result.get(UpdateResult.ERRORED.name)[user._id] = user_result.get(UpdateResult.ERRORED.name)
+
+        # Output update results to console
+        logger.info(f'{UpdateResult.SUCCEEDED.name} = {update_result.get(UpdateResult.SUCCEEDED.name)}')
+        logger.info(f'{UpdateResult.SKIPPED.name} = {update_result.get(UpdateResult.SKIPPED.name)}')
+        logger.warning(f'{UpdateResult.FAILED.name} = {update_result.get(UpdateResult.FAILED.name)}')
+        logger.error(f'{UpdateResult.ERRORED.name} = {update_result.get(UpdateResult.ERRORED.name)}')
+        if dry_run:
+            logger.warning(f'The above output is the result from a dry-run pass! '
+                           f'{UpdateResult.SUCCEEDED.name} ones were not actually added! '
+                           f'{UpdateResult.ERRORED.name} is expected to be an empty list!')
+
+
+def update_email_domain(user, src_domain, dst_domain, dry_run=True):
+    """For a given user, if it has an email `example@<src_domain>`, attempt to add `example@<dst_domain>` to this
+    account. The action will be skipped if `example@<dst_domain>` already exists on this user or no email is found
+    under `@<src_domain>`. The action will fail if `example@<dst_domain>` already exists on another account. Other
+    unexpected exceptions will be considered as error.
+    """
+
+    # Find eligible emails to add
+    emails_to_add = []
+    user_result = {
+        UpdateResult.SUCCEEDED.name: [],
+        UpdateResult.SKIPPED.name: [],
+        UpdateResult.FAILED.name: [],
+        UpdateResult.ERRORED.name: [],
+    }
+    for email in user.emails.filter(address__endswith=f'@{src_domain}'):
+        email_parts = email.address.split('@')
+        name_part = email_parts[0].lower()
+        domain_part = email_parts[1].lower()
+        if domain_part == src_domain:
+            emails_to_add.append(f'{name_part}@{dst_domain}')
+    if not emails_to_add:
+        logger.warning(f'Action skipped due to no eligible email found for user [{user._id}]!')
+        return user_result
+    # Verify and attempt to add email; keep track of successes, failures and errors
+    for email in emails_to_add:
+        duplicate_user = get_user(email=email)
+        if not duplicate_user:
+            try:
+                if not dry_run:
+                    user.emails.create(address=email)
+            except Exception:
+                logger.error(f'An unexpected error occurred when adding email [{email}] to user [{user._id}]!')
+                sentry.log_exception()
+                user_result.get(UpdateResult.ERRORED.name).append(email)
+            else:
+                logger.info(f'Successfully added email [{email}] to user [{user._id}]!')
+                user_result.get(UpdateResult.SUCCEEDED.name).append(email)
```

Comment:
I decided to log "success" and update result map even it is a dry run so that you can actually see the emails to be added in the output. The differences between dry run and actual run are:

* Emails are not added
* Error list will be always be empty since unexpected exception can't happen when emails are not added